### DISCUSSION TITLE
Usage of Tinfour in Geometry Suite

### DISCUSSION THREAD

**Discussion by micycle1** [gwlucastrig/Tinfour#74] [Show and tell]
I've just finished version 1.0.0 of [my geometry suite](https://github.com/micycle1/PGS) for [Processing](https://processing.org/).

I thought I'd share where I've used Tinfour within the project:

### Voronoi
![voronoiCircles](https://user-images.githubusercontent.com/9304234/117369094-551d4b80-aebc-11eb-9440-1a140726ab7b.gif)
![voronoi](https://user-images.githubusercontent.com/9304234/117369116-5d758680-aebc-11eb-941d-0603bd3dc600.gif)

### Isolines/Contours
![isolines](https://user-images.githubusercontent.com/9304234/117369098-564e7880-aebc-11eb-89b4-458ad599a78a.gif)
![isolines2](https://user-images.githubusercontent.com/9304234/117369102-58183c00-aebc-11eb-8938-6c720c3acce0.gif)

### Medial Axis
![medialAxis](https://user-images.githubusercontent.com/9304234/117369108-5a7a9600-aebc-11eb-8b57-452c176d1fdc.gif)

### Triangulation
![poissonTriangulation](https://user-images.githubusercontent.com/9304234/117369111-5b132c80-aebc-11eb-824e-664819303c73.gif)

### Triangulation w/ refinement
<img width="500" alt="triangulation2" src="https://user-images.githubusercontent.com/9304234/117369561-fa382400-aebc-11eb-829c-dcc17a88f298.png">

**Comment by gwlucastrig** [comment]
Wow.  Just Wow...

**Comment by gwlucastrig** [comment]
I am particularly impressed with the Medial Axis and Triangulation refinement images. These are problems I've worked on myself. So I know from direct experience that both of them are challenging.  I particularly impressed by how well the Medial Axis propagates so cleanly into the various segments of the maple leaf.

Delaunay refinement is on the Tinfour roadmap (it is, in fact, the last problem remaining before Tinfour is "done").   I will be looking at your code to see what kind of insights it provides.

**Reply by micycle1** [reply]
Thanks haha.

I prune the medial axis segments based on underlying feature area (the area of triangles that underlie a segment) and axial gradient.

![image](https://user-images.githubusercontent.com/9304234/117453358-c43a8480-af3c-11eb-8c35-3be798a5b8c5.png)


I went for a very dumb approach to refinement: (recursively) inserting triangle centroids as vertices into the Tin; I found that centroids worked better than circumcenters. I did look at Ruppert refinement but that can wait for now.
